### DISCUSSION TITLE
Use the @ViewAction macro for a generic Reducer

### DISCUSSION THREAD

**Discussion by pomozoff** [pointfreeco/swift-composable-architecture#2818] [Q&A]
I've a generic reducer defined like as following:
```swift
@Reducer
public struct Feature<T: Taggable> {
...
```
I'd like to use view actions separately from domain actions with the macro `@ViewAction(for:)`, but I don't know how to configure the generic for the Feature in my View:
```swift
@ViewAction(for: Feature<???>.self)
public struct TagsScrollView<T: Taggable>: View {
  public var body: some View {
...
```
Is it a good practice to use generic reducers in TCA?

**Comment by mbrandonw** [comment]
Hi @pomozoff, generic reducers are completely fine in TCA, but you cannot use the `@ViewAction` macro with them.

To be honest, generic reducers seem to be a tool for building library code that are integrated into app features, rather than app features themselves. And so from that perspective "view actions" don't actually seem to be that useful. 

However, if you really want to use view actions, then you can just implement what `@ViewAction` is doing for you manually. You can even look at the tests to see what it expands to (it's not much):

https://github.com/pointfreeco/swift-composable-architecture/blob/af38644542b1310c0bb023a5af2545beb8854009/Tests/ComposableArchitectureMacrosTests/ViewActionMacroTests.swift#L17-L39

The only thing you can't recover that the macro provides is the lint warnings that let you know when you are sending an action directly into the store:

https://github.com/pointfreeco/swift-composable-architecture/blob/af38644542b1310c0bb023a5af2545beb8854009/Tests/ComposableArchitectureMacrosTests/ViewActionMacroTests.swift#L285-L287

But again, since generic reducers tend to be library code, I don't think that's a huge deal.

**Reply by pomozoff** [reply]
Hi @mbrandonw!

Thank you, looks like I got your point. Will try to do the wrapper manually.

Best regards,
Anton
